src/transform/activation.jl

```diff
@@ -0,0 +1,132 @@
+# Repurposed from https://github.com/tbreloff/OnlineAI.jl
+# -------------------------------------------------------------
+
+abstract Activation <: Transformation
+
+value(activation::Activation, input::AbstractVector) = Float64[value(activation, Σ) for Σ in input]
+deriv(activation::Activation, input::AbstractVector) = Float64[deriv(activation, Σ) for Σ in input]
+
+function value!(buffer::AbstractVector, activation::Activation, input::AbstractVector)
+    for i in 1:length(buffer)
+        buffer[i] = value(activation, input[i])
+    end
+end
+function deriv!(buffer::AbstractVector, activation::Activation, input::AbstractVector)
+    for i in 1:length(buffer)
+        buffer[i] = deriv(activation, input[i])
+    end
+end
+
+# ----------------------------------------------------------------------------
+
+doc"""
+f(x) = x
+f'(x) = 1
+"""
+immutable IdentityActivation <: Activation end
+value(activation::IdentityActivation, input::Real) = input
+deriv(activation::IdentityActivation, input::Real) = 1.0
+
+# ----------------------------------------------------------------------------
+
+doc"""
+f(x) = 1 / (1 + exp(-x))
+f'(x) = f(x) * (1 - f(x))
+"""
+immutable SigmoidActivation <: Activation end
+value(activation::SigmoidActivation, input::Real) = 1.0 / (1.0 + exp(-input))
+deriv(activation::SigmoidActivation, input::Real) = (s = value(activation, input); s * (1.0 - s))
+
+# ----------------------------------------------------------------------------
+
+doc"""
+f(x) = tanh(x)
+f'(x) = 1 - tanh(x)²
+"""
+immutable TanhActivation <: Activation end
+value(activation::TanhActivation, input::Real) = tanh(input)
+deriv(activation::TanhActivation, input::Real) = 1.0 - tanh(input)^2
+
+# ----------------------------------------------------------------------------
+
+doc"""
+f(x) = x / (1 + |x|)
+f'(x) = 1 / (1 + |x|)²
+"""
+immutable SoftsignActivation <: Activation end
+value(activation::SoftsignActivation, input::Real) = input / (1.0 + abs(input))
+deriv(activation::SoftsignActivation, input::Real) = 1.0 / (1.0 + abs(input))^2
+
+# ----------------------------------------------------------------------------
+
+# Rectified Linear Unit
+doc"""
+f(x) = max(0, x)
+f'(x) = { 1,  x > 0
+        { 0,  x ≤ 0
+"""
+immutable ReLUActivation <: Activation end
+value(activation::ReLUActivation, input::Real) = max(0.0, input)
+deriv(activation::ReLUActivation, input::Real) = float(input >= 0.0)
+
+# ----------------------------------------------------------------------------
+
+# Leaky Rectified Linear Unit: modified derivative to fix "dying ReLU" problem
+doc"""
+f(x) = max(0, x)
+f'(x) = { 1,  x > 0
+        { ρ,  x ≤ 0
+"""
+immutable LReLUActivation <: Activation
+    ρ::Float64
+end
+LReLUActivation() = LReLUActivation(0.01)
+value(activation::LReLUActivation, input::Real) = deriv(activation, input) * input 
+deriv{T<:Real}(activation::LReLUActivation, input::T) = input >= zero(T) ? one(T) : activation.ρ
```

Comment:
This is tricky. Although it looks like it, this function is not type stable. That is because `activation.p` is always a `Float64`, while `T` can be any `Real`. This will result in slow code if input is anything but `Float64`.

I would suggest adding a type parameter to `LReLUActivation`

```julia
immutable LReLUActivation{T<:Real} <: Activation
    ρ::T
end
LReLUActivation{T}(p::T = 0.01) = LReLUActivation{T}(p)
```

and then changing this line to

```julia
deriv{T<:Real}(activation::LReLUActivation{T}, input::T) = input >= zero(T) ? one(T) : activation.ρ
```

or the more general one which allows for different input types

```julia
function deriv{T<:Real, R<:Real}(activation::LReLUActivation{T}, input::R)
    N = promote_type(T, R)
    input >= 0 ? one(N) : N(activation.ρ)
end
```